download/download.go

```diff
@@ -0,0 +1,119 @@
+package download
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+
+	"github.com/ONSdigital/dp-api-clients-go/v2/clientlog"
+	dperrors "github.com/ONSdigital/dp-api-clients-go/v2/errors"
+	healthcheck "github.com/ONSdigital/dp-api-clients-go/v2/health"
+	health "github.com/ONSdigital/dp-healthcheck/healthcheck"
+	dprequest "github.com/ONSdigital/dp-net/request"
+	"github.com/ONSdigital/log.go/v2/log"
+)
+
+const service = "download-service"
+
+// Client is an download service client which can be used to make requests to the server.
+// It extends the generic healthcheck Client structure.
+type Client struct {
+	hcCli *healthcheck.Client
+}
+
+// closeResponseBody closes the response body and logs an error if unsuccessful
+func closeResponseBody(ctx context.Context, resp *http.Response) {
+	if resp.Body != nil {
+		if err := resp.Body.Close(); err != nil {
+			log.Error(ctx, "error closing http response body", err)
+		}
+	}
+}
+
+// NewAPIClient creates a new instance of DownloadServiceAPI Client with a given download service url
+func NewAPIClient(downloadServiceAPIURL string) *Client {
+	return &Client{
+		healthcheck.NewClient(service, downloadServiceAPIURL),
+	}
+}
+
+// NewWithHealthClient creates a new instance of DownloadServiceAPI Client,
+// reusing the URL and Clienter from the provided healthcheck client.
+func NewWithHealthClient(hcCli *healthcheck.Client) *Client {
+	return &Client{
+		healthcheck.NewClientWithClienter(service, hcCli.URL, hcCli.Client),
+	}
+}
+
+// URL returns the URL used by this client
+func (c *Client) URL() string {
+	return c.hcCli.URL
+}
+
+// HealthClient returns the underlying Healthcheck Client for this download service client
+func (c *Client) HealthClient() *healthcheck.Client {
+	return c.hcCli
+}
+
+// Checker calls download service health endpoint and returns a check object to the caller.
+func (c *Client) Checker(ctx context.Context, check *health.CheckState) error {
+	return c.hcCli.Checker(ctx, check)
+}
+
+// Download returns the requested path
+func (c *Client) Download(ctx context.Context, userAuthToken, serviceAuthToken, path string) (*Response, error) {
+	uri := fmt.Sprintf("%s/downloads-new/%s", c.hcCli.URL, path)
+
+	clientlog.Do(ctx, "retrieving resource", service, uri)
+
+	resp, err := c.doGetWithAuthHeaders(ctx, userAuthToken, serviceAuthToken, uri, nil)
+	if err != nil {
+		return nil, dperrors.New(
+			fmt.Errorf("failed to create request to DownloadService API: %s", err),
+			http.StatusInternalServerError,
+			nil,
+		)
+	}
+	defer closeResponseBody(ctx, resp)
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, dperrors.New(
+			fmt.Errorf("failed to read error response body: %s", err),
+			resp.StatusCode,
+			nil,
+		)
+	}
+
+	if resp.StatusCode == http.StatusMovedPermanently {
+		return &Response{RedirectUrl: resp.Header.Get("location")}, nil
```

Comment:
Could you resolve the redirect and just return a ReaderCloser from this function rather than a Response object.

Reply:
👍  i think good to know its a redirect though so am thinking to keep `RedirectUrl` in response - can just ignore. just thinking for us - we will persist this and stop the redirect next time. But we will prob persist on publish anyway so never hit this scenario 🙈

Reply:
forget this - httpclient by default follows redirect 😄